packages/static-build/src/utils/build-output-v2.ts

```diff
@@ -0,0 +1,124 @@
+import path from 'path';
+import { stat, pathExists } from 'fs-extra';
+import { Route } from '@vercel/routing-utils';
+import {
+  Files,
+  FileFsRef,
+  debug,
+  glob,
+  EdgeFunction,
+} from '@vercel/build-utils';
+import { isObjectEmpty } from './_shared';
+
+const BUILD_OUTPUT_DIR = '.output';
+
+/**
+ * Returns the path to the Build Output API v2 directory when the
+ * `config.json` file was created by the framework / build script,
+ * or `undefined` if the framework did not create the v3 output.
+ */
+export async function getBuildOutputDirectory(
+  workingDir: string
+): Promise<string | undefined> {
+  const outputDir = path.join(workingDir, BUILD_OUTPUT_DIR);
+  const outputPathExists = await pathExists(outputDir);
+  if (outputPathExists) {
+    return outputDir;
+  }
+  return undefined;
+}
+
+/**
+ * Reads the BUILD_OUTPUT_DIR directory and returns and object
+ * that should be merged with the build outputs.
+ */
+export async function readBuildOutputDirectory({
+  workPath,
+}: {
+  workPath: string;
+}) {
+  // Functions are not supported, but are used to support Middleware
+  const functions: Record<string, EdgeFunction> = {};
+
+  // Routes are not supported, but are used to support Middleware
+  const routes: Array<Route> = [];
+
+  const middleware = await getMiddleware(workPath);
+  if (middleware) {
+    routes.push(middleware.route);
+
+    functions['middleware'] = new EdgeFunction({
+      deploymentTarget: 'v8-worker',
+      entrypoint: '_middleware.js',
+      files: {
+        '_middleware.js': middleware.file,
+      },
+      name: 'middleware',
+    });
+  }
+
+  const staticFiles = await readStaticFiles({ workPath });
+
+  const outputs = {
+    staticFiles: isObjectEmpty(staticFiles) ? null : staticFiles,
+    functions: isObjectEmpty(functions) ? null : functions,
+    routes: routes.length ? routes : null,
+  };
+
+  if (outputs.functions) {
+    debug(`Detected Serverless Functions in "${BUILD_OUTPUT_DIR}"`);
+  }
+
+  if (outputs.staticFiles) {
+    debug(`Detected Static Assets in "${BUILD_OUTPUT_DIR}"`);
+  }
+
+  if (outputs.routes) {
+    debug(`Detected Routes Configuration in "${BUILD_OUTPUT_DIR}"`);
+  }
+
+  return outputs;
+}
+
+async function getMiddleware(
+  workPath: string
+): Promise<{ route: Route; file: FileFsRef } | undefined> {
+  const middlewareRelativePath = path.join(
```

Comment:
Instead of assuming this file, let's read the `.output/functions-manifest.json` file as the source of truth (middleware has `runtime: 'web'`.